sphinx build errors

steenzout opened the pull request:
Hello,

  this PR is to make sphinx doc generation warnings and link check break the build if something gets changed on the docs that isn't right.

  I had to:
  - change the import for sqlalchemy in the `luigi/db_task_history.py` so that sphinx wouldn't complain (hope it doesn't look that ugly!)
  - fix all the warnings on the docstrings
  - fix URLs so that permanent redirects don't throw sphinx errors anymore
  - had to make a copy of the README.rst into `doc/readme.rst` since links to external images / figures will throw a sphinx warning during the documentation build (is this ok? at least, for now?)
  - 2 minor sphinx configuration tweaks to fix / bypass / overcome sphinx warnings
  - added docs-test tox environment
  - added docs-test tox environment to the travis build

  I also added a few changes to the .gitignore.
Based on the content you previously had I:
  - generated new rules on gitignore.io and added them.
  - removed what became duplicate from the top of the file
  - ordered the project custom rules by alphabetical order
  - added a few python files that show up after I run the tests

  let me know if you want me to rebase all of this into 1 commit (some projects don't like a series of commits).

  comments welcomed!...

cheers

themalkolm commented on tox.ini:
```diff
@@ -31,10 +31,21 @@ deps = autopep8
   sphinx_rtd_theme
 changedir = doc
-commands = /usr/bin/make {posargs}
+commands = /usr/bin/make SPHINXOPTS='-W' {posargs}
```
What if we use sphinx-build instead? https://github.com/spotify/luigi/pull/697

steenzout replied:
I saw the PR.
either way works.
I'll wait for the merge for #697 and then I'll rebase.

themalkolm commented on doc/readme.rst:
```diff
@@ -0,0 +1,134 @@
+.. figure:: luigi.png
```
So now doc/readme.rst and readme.rst in the root go out of sync easily?

steenzout replied:
they can, yes.
I don't see an easy alternative to have both in sync though due to the external image problem though.

maybe they actually should be different things and have different content?
simplify the README github page (ala https://github.com/ansible/ansible)?
make a more in-depth intro on the readthedocs (just like it is)?

erikbern commented:
Looks awesome! Except
1. Not great with two README files that go out of sync
2. Can you squash a bunch of the commits? I prefer it to be just a handful

steenzout commented:
@erikbern changes squash for 2.

regarding 1. I've added a note.
I'm not sure how to not dup it.

erikbern commented:
I don't think we need to syntax check README.rst since you can look at it by just going to the branch on Github. Can we just avoid checking it?

themalkolm commented:
@steenzout Maybe add a copy command in [docs] to copy README.rst to doc before docs being run? :) No duplication, everybody is happy.

erikbern commented:
+1

steenzout commented:
@themalkolm if I copy it, sphinx will try to validate it when doing HTML docs generation just like when it was pointing to the ../README.rst. :)

I can add a command to copy it then and perform a cleanup but next time some adds something to the main README.rst it may break the docs generation.
I guess that's ok.

give me 1h...

themalkolm commented:
I would say it is OK to break documentation if README.rst is broken.

themalkolm commented:
Maybe adding [docs] to default targets to tox would ensure docs are never borked.

Tarrasch commented:
This commit message is unreadable since the header does not describe the overall change. The imporpant change of this patch set is not that you ignore E308.

steenzout commented:
@Tarrasch I've changed the 1st line of the commit message.
hope this is what you wanted or else just let me know.

@themalkolm I've merged the docsapi tox env with the docs tox env since it didn't made any sense to have it separated (I didn't realize that you had to build the docs in 2 steps: api and then HTML).
I've noticed that due to this a luigi.tools.rst file was generated by sphinx.
since it wasn't being included it complained and I fixed it.
please check if on the API Reference it's where you wanted it to be.

to sync the README.rst I copy the file over, change github URL to local file references and remove the badges using sed commands.
sphinx is now happy.

if you really want the badges on the docs (?) we'd have to copy the badges every time we try to generate the documentation and be sure you're picking up the badge for the right branch / tag.

erikbern commented:
Great! Maybe just squash everything since the first commit is already a monster-commit

steenzout commented:
squash done.
made sed compatible for ubuntu and macosx.

erikbern commented on tox.ini:
```diff
@@ -30,13 +30,34 @@ commands = pep8 --ignore E501 luigi test examples bin
+    sphinx-apidoc -o doc/api -T luigi
+
+    # sync README.rst file
+    cp README.rst doc/README.rst
```
Can you delete doc/readme.rst since you're copying README.rst into doc/README.rst now?

Also Sphinx is throwing some errors. Could it be because you need to copy README.rst into doc/readme.rst (lower case) rather than doc/README.rst?

```Warning, treated as error:
/home/travis/build/spotify/luigi/doc/readme.rst:: WARNING: document isn't included in any toctree```

steenzout commented:
@erikbern I saw that.
It worked on my laptop but I was missing a commit to delete the readme.rst (lowercase) file.
I've fixed that now.
it should fix both errors.

sorry about the extra spam!

erikbern commented:
awesome!

erikbern commented:
this was commit #1,500 :)

steenzout commented:
:)
